docs/storage.md

```diff
@@ -449,6 +450,112 @@ config:
   directory: ""
 ```
 
+### OCI
+
+[Oracle Cloud Infrastructure](https://www.oracle.com/cloud). To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
+You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.
+More information about [Overview of Object Storage](https://docs.cloud.oracle.com/en-us/iaas/Content/Object/Concepts/objectstorageoverview.htm).
+
+OCI Object Storage supports three different providers, default, instance principle, and raw. 
+The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.
+
+By default, like other clients, Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. 
+This means each bucket can only store one prometheus instance data.
```

Comment:
Sorry for the poor description.
When thanos uploads all the blocks from a prometheus instance, it will create a directory structure similar to the following in OCI objectstore
```text
|-- 01FF8QX35297NG9HCDHGH1BN1
|   |-- chunks
|       |-- 000001
|       |-- 000002
|       |-- ....
|-- 01FFPFKMAQQNF4N945X7JWPQ8
|   |-- chunks
|.      |-- ....
|-- .... other prometheus blocks
|
|-- debug
    |-- meta
        |-- .....
```
`I consider above as one prometheus instance data.`

In this PR, a new option called object_base_path is added.
If defined, the whole directory structure above will be uploaded under the object_base_path prefix.

In second screenshot in this PR, it holds 4 prometheus instances.
Prometheus A - object_base_path is set to thanos.dev.thanos
Prometheus B - object_base_path is set to thanos.meta.sa/0
Prometheus C - object_base_path is set to thanos.meta.sa/1
Prometheus D - object_base_path is set to thanos.meta.sa/2

Here is the illustration in OCI objectstorage bucket.
```
|-- thanos.dev.thanos
    |--  <<Promethus-A with all blocks and debug meta>>
|-- thanos.meta.sa
    |-- 0 <<Promethus-B with all blocks and debug meta>>
    |-- 1 <<Promethus-C with all blocks and debug meta>>
    |-- 2 <<Promethus-D with all blocks and debug meta>>
        |-- 01FF8QX35297NG9HCDHGH1BN1
        |   |-- chunks
        |       |-- 000001
        |       |-- 000002
        |       |-- ....
        |-- 01FFPFKMAQQNF4N945X7JWPQ8
        |   |-- chunks
        |.      |-- ....
        |-- .... other prometheus blocks
        |
        |-- debug
            |-- meta
                |-- .....
```

If you want to provide a better description for the Thanos community, I will be happy to change it in my PR.

Reply:
Yes, that's what I mean - this https://github.com/thanos-io/thanos/issues/1318 has been a long-standing issue. I think this kind of functionality needs to be added "one level above" i.e. not at the raw object storage calls level. Otherwise, it would be weird to have duplicate functionality. I think others would agree as well. My suggestion would be to remove this prefix functionality and move it one level above so that it could be used with any kind of object storage. It could be done in a separate pull request. We could double check in #thanos-dev, though

Reply:
Got it. Yes, adding the feature to `one level up` is a great idea for all Objectstorage clients.

Since allowing the same bucket to hold multiple prometheus instances' data is the main use case of our team and we would like to use Thanos master branch instead of my forked branch.

I can think of three different approaches (order by my preferences) to avoid duplicate functionality in my PR:
**_Approach # 1_**
- hide the object_base_path configuration from storage.md documentation
- without the object_base_path setting, my PR will act exactly the same as other ObjecctStorage Clients
- the hidden feature can still be used by the insider who know the existence of this feature
- after the code check it, whoever want to work on #1318 will have an example how OCI client is implemented.

**_Approach # 2_**
- hide the object_base_path configuration from storage.md documentation
- hard code the object_base_path to empty string and won't accept any hidden customization

**_Approach # 3_**
Remove all the object_base_path throughout the PR.


Please let me know which approach is acceptable for you.
Thanks!

Reply:
Sorry for the delay again, bit busy times. 
As I like the different approaches I'm not that fan of hidden features in a way that one needs to know the existence in an open source project. Also I took some time to read the long standing issue and the consensus after the discussion was this:

> Looks like the decision is that we are ok for simple custom prefix on config level (for all objstorages) in YAML (:

So, I would say that if we implement it, it should be "that". Still open for discussion though, but then it should require a lot more discussion from everyone else again IMO ^^

I would say, let's just implement this on level higher as discussed in that main issue. For all the storage backends, implement at least the config in a way that it can say "This is not supported yet for this object store" if one would configure it. With then the exception for OCI, in which you did implement it. 

This gives a bit more motivation to get that issue fixed, and we can have iterations per objectstorage to actually implement it. I think by saying "It needs to be done for them all in one go", is really limiting us to get it moving forward. 

What do you think? Also, I'm actually happy to help on that. I got some spare time soon.

Reply:
I really think we should do (3) approach here. Then, you could implement it in a follow up PR. We have interfaces for object storage so IMHO it should be fairly easy to implement prefix functionality.

Reply:
I will remove the prefix option when I have a chance.

Reply:
Don't forget to remove this sentence as it is not true in practice, it is _very_ hard to get a UUID collision.